In [1]:
import tensorflow as tf
tf.test.gpu_device_name()

'/device:GPU:0'

In [4]:
!pip uninstall torch -y


Uninstalling torch-0.3.1:
  Successfully uninstalled torch-0.3.1


In [4]:
!pip install http://download.pytorch.org/whl/cu80/torch-0.3.1-cp36-cp36m-linux_x86_64.whl  && pip install torchvision

In [5]:
!pip install Pillow==4.0.0
!pip install PIL
!pip install image

    100% |████████████████████████████████| 5.6MB 6.0MB/s 
torchvision 0.2.1 has requirement pillow>=4.1.1, but you'll have pillow 4.0.0 which is incompatible.
  Found existing installation: Pillow 5.3.0
    Uninstalling Pillow-5.3.0:
      Successfully uninstalled Pillow-5.3.0
  Could not find a version that satisfies the requirement PIL (from versions: )
No matching distribution found for PIL
    100% |████████████████████████████████| 7.3MB 5.6MB/s 


In [8]:
!pip install git+https://github.com/fastai/fastai@2e1ccb58121dc648751e2109fc0fbf6925aa8887

  Cloning https://github.com/fastai/fastai (to revision 2e1ccb58121dc648751e2109fc0fbf6925aa8887) to /tmp/pip-req-build-oaoowhch
  Running setup.py bdist_wheel for fastai ... - \ done
  Stored in directory: /tmp/pip-ephem-wheel-cache-pro9lpot/wheels/25/e9/9b/223e47f59c39f0bd393733db0f7efe8ec6f078d29c32198f9f
Successfully built fastai


In [9]:
!apt update && apt install -y libsm6 libxext6

Ign:1 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1710/x86_64  InRelease
Hit:2 http://ppa.launchpad.net/graphics-drivers/ppa/ubuntu bionic InRelease
Hit:3 http://archive.ubuntu.com/ubuntu bionic InRelease
Get:4 http://security.ubuntu.com/ubuntu bionic-security InRelease [83.2 kB]
Ign:5 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64  InRelease
Hit:6 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1710/x86_64  Release
Hit:7 https://developer.download.nvidia.com/compute/machine-learning/repos/ubuntu1604/x86_64  Release
Get:8 http://archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:9 http://archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Get:12 http://security.ubuntu.com/ubuntu bionic-security/universe amd64 Packages [104 kB]
Get:13 http://archive.ubuntu.com/ubuntu bionic-updates/universe Sources [115 kB]
Get:14 http://security.ubuntu.com/ubuntu bionic-security/main amd64 Packages [229 k

In [0]:
from fastai.imports import *
from fastai.structured import *
from pandas_summary import DataFrameSummary
from sklearn.ensemble import RandomForestRegressor, RandomForestClassifier
from IPython.display import display
from sklearn import metrics

In [12]:
!pip install kaggle

    100% |████████████████████████████████| 61kB 2.8MB/s 
    100% |████████████████████████████████| 235kB 6.0MB/s 
  Running setup.py bdist_wheel for kaggle ... - \ done
  Stored in directory: /root/.cache/pip/wheels/44/2c/df/22a6eeb780c36c28190faef6252b739fdc47145fd87a6642d4
  Running setup.py bdist_wheel for python-slugify ... - done
  Stored in directory: /root/.cache/pip/wheels/e3/65/da/2045deea3098ed7471eca0e2460cfbd3fdfe8c1d6fa6fcac92
Successfully built kaggle python-slugify


In [13]:
!pwd

/content


In [14]:
!ls -a

.  ..  .config	sample_data


In [0]:
! rm -rf kaggle


In [67]:
!ls -a

.  ..  .config	sample_data


In [0]:
! rm -rf kaggle


In [0]:
!mkdir /content/.kaggle 

I had problems with .kaggle. I got the error when i used !mkdir ~/.kaggle. I *FileNotFoundError: [Errno 2] No such file or directory: '/content/.kaggle/kaggle.json'*

When i did !ls -a i couldnt see the .kaggle folder under /content. I fixed it by using this command - !mkdir /content/.kaggle  

cp: cannot create regular file '/root/.kaggle/kaggle.json': No such file or directory

In [3]:
!ls -a

.  ..  bulldozers  .config  .kaggle  sample_data


In [0]:
import json


In [0]:
import zipfile
import os
api_token = {"username":"bhuvanaka","key":"a003c789c41934e7eb90922841a4009e"}
with open('/content/.kaggle/kaggle.json', 'w') as file:
    json.dump(api_token, file)

In [0]:
!chmod 600 /content/.kaggle/kaggle.json


In [0]:
!mkdir ~/.kaggle

This code fixes the following error - cp: cannot create regular file '/root/.kaggle/kaggle.json': No such file or directory :

In [0]:
!cp /content/.kaggle/kaggle.json ~/.kaggle/kaggle.json


In [0]:
!mkdir bulldozers


In [24]:
!kaggle competitions download -c bluebook-for-bulldozers -p /bulldozers/ 

 73% 5.00M/6.85M [00:00<00:00, 38.0MB/s]
100% 6.85M/6.85M [00:00<00:00, 49.2MB/s]
 54% 5.00M/9.28M [00:00<00:00, 44.7MB/s]
100% 9.28M/9.28M [00:00<00:00, 72.4MB/s]
  0% 0.00/209k [00:00<?, ?B/s]
100% 209k/209k [00:00<00:00, 144MB/s]
  0% 0.00/3.17M [00:00<?, ?B/s]
100% 3.17M/3.17M [00:00<00:00, 253MB/s]
  0% 0.00/297k [00:00<?, ?B/s]
100% 297k/297k [00:00<00:00, 125MB/s]
  0% 0.00/10.8k [00:00<?, ?B/s]
100% 10.8k/10.8k [00:00<00:00, 12.8MB/s]
  0% 0.00/192k [00:00<?, ?B/s]
100% 192k/192k [00:00<00:00, 106MB/s]
 71% 35.0M/49.1M [00:00<00:00, 60.0MB/s]
100% 49.1M/49.1M [00:00<00:00, 132MB/s] 
  0% 0.00/316k [00:00<?, ?B/s]
100% 316k/316k [00:00<00:00, 178MB/s]
 71% 5.00M/7.06M [00:00<00:00, 39.9MB/s]
100% 7.06M/7.06M [00:00<00:00, 53.3MB/s]
 95% 108M/114M [00:00<00:00, 117MB/s] 
100% 114M/114M [00:00<00:00, 175MB/s]
 52% 5.00M/9.59M [00:00<00:00, 46.1MB/s]
100% 9.59M/9.59M [00:00<00:00, 75.9MB/s]
  0% 0.00/3.40M [00:00<?, ?B/s]
100% 3.40M/3.40M [00:00<00:00, 256MB/s]
  0% 0.00/207k [00:0

In [0]:
os.chdir('/bulldozers/')


In [11]:
!pwd


/bulldozers


In [12]:
!ls

Data%20Dictionary.xlsx		  Train.7z	     Valid.7z
Machine_Appendix.csv		  TrainAndValid.7z   Valid.csv
median_benchmark.csv		  TrainAndValid.csv  ValidSolution.csv
random_forest_benchmark_test.csv  TrainAndValid.zip  Valid.zip
Test.csv			  Train.csv
tmp				  Train.zip


In [0]:
!unzip -q Train.zip

In [0]:
df_raw = pd.read_csv('/bulldozers/Train.csv', low_memory=False, 
                     parse_dates=["saledate"])


In [15]:
def display_all(df):
    with pd.option_context("display.max_rows", 1000): 
        with pd.option_context("display.max_columns", 1000): 
            display(df)
display_all(df_raw.tail().transpose())

401120  \
SalesID                                                             6333336   
SalePrice                                                             10500   
MachineID                                                           1840702   
ModelID                                                               21439   
datasource                                                              149   
auctioneerID                                                              1   
YearMade                                                               2005   
MachineHoursCurrentMeter                                                NaN   
UsageBand                                                               NaN   
saledate                                                2011-11-02 00:00:00   
fiModelDesc                                                           35NX2   
fiBaseModel                                                              35   
fiSecondaryDesc                                                          NX   
fiModelSeries                                                             2   
fiModelDescriptor                                                       NaN   
ProductSize                                                            Mini   
fiProductClassDesc        Hydraulic Excavator, Track - 3.0 to 4.0 Metric...   
state                                                              Maryland   
ProductGroup                                                            TEX   
ProductGroupDesc                                           Track Excavators   
Drive_System                                                            NaN   
Enclosure                                                             EROPS   
Forks                                                                   NaN   
Pad_Type                                                                NaN   
Ride_Control                                                            NaN   
Stick                                                                   NaN   
Transmission                                                            NaN   
Turbocharged                                                            NaN   
Blade_Extension                                                         NaN   
Blade_Width                                                             NaN   
Enclosure_Type                                                          NaN   
Engine_Horsepower                                                       NaN   
Hydraulics                                                        Auxiliary   
Pushblock                                                               NaN   
Ripper                                                                  NaN   
Scarifier                                                               NaN   
Tip_Control                                                             NaN   
Tire_Size                                                               NaN   
Coupler                                                 None or Unspecified   
Coupler_System                                                          NaN   
Grouser_Tracks                                                          NaN   
Hydraulics_Flow                                                         NaN   
Track_Type                                                            Steel   
Undercarriage_Pad_Width                                 None or Unspecified   
Stick_Length                                            None or Unspecified   
Thumb                                                   None or Unspecified   
Pattern_Changer                                         None or Unspecified   
Grouser_Type                                                         Double   
Backhoe_Mounting                                                        NaN   
Blade_Type                                                              NaN   
Travel_Controls                         

In [19]:
df_raw.SalePrice

0         66000
1         57000
2         10000
3         38500
4         11000
5         26500
6         21000
7         27000
8         21500
9         65000
10        24000
11        22500
12        36000
13        30500
14        28000
15        19000
16        13500
17         9500
18        12500
19        11500
20        41000
21        34500
22        26000
23        73000
24        33000
25        12500
26        15500
27        53000
28        46000
29        89000
          ...  
401095    10500
401096    10000
401097     8500
401098    10500
401099    11000
401100    10500
401101    10000
401102    10500
401103    12500
401104    10500
401105    10000
401106    10500
401107    12500
401108    10500
401109    10500
401110    10500
401111    10000
401112    10500
401113    10500
401114    12500
401115    10000
401116     8500
401117    10000
401118     9500
401119     8500
401120    10500
401121    11000
401122    11500
401123     9000
401124     7750
Name: SalePrice, Length:

In [0]:
df_raw.SalePrice=np.log(df_raw.SalePrice)

In [34]:
df_raw.SalePrice

0         11.097410
1         10.950807
2          9.210340
3         10.558414
4          9.305651
5         10.184900
6          9.952278
7         10.203592
8          9.975808
9         11.082143
10        10.085809
11        10.021271
12        10.491274
13        10.325482
14        10.239960
15         9.852194
16         9.510445
17         9.159047
18         9.433484
19         9.350102
20        10.621327
21        10.448715
22        10.165852
23        11.198215
24        10.404263
25         9.433484
26         9.648595
27        10.878047
28        10.736397
29        11.396392
            ...    
401095     9.259131
401096     9.210340
401097     9.047821
401098     9.259131
401099     9.305651
401100     9.259131
401101     9.210340
401102     9.259131
401103     9.433484
401104     9.259131
401105     9.210340
401106     9.259131
401107     9.433484
401108     9.259131
401109     9.259131
401110     9.259131
401111     9.210340
401112     9.259131
401113     9.259131


In [0]:
m=RandomForestRegressor(n_jobs=-1)

In [23]:
m.fit(df_raw.drop('SalePrice',axis =1),df_raw.SalePrice)

ValueError: ignored

In [18]:
df_raw.saledate

0        2006-11-16
1        2004-03-26
2        2004-02-26
3        2011-05-19
4        2009-07-23
5        2008-12-18
6        2004-08-26
7        2005-11-17
8        2009-08-27
9        2007-08-09
10       2008-08-21
11       2006-08-24
12       2005-10-20
13       2006-01-26
14       2006-01-03
15       2006-11-16
16       2007-06-14
17       2010-01-28
18       2006-03-09
19       2005-11-17
20       2006-05-18
21       2006-10-19
22       2007-10-25
23       2006-10-19
24       2004-05-20
25       2006-03-09
26       2006-03-09
27       2007-02-22
28       2007-08-09
29       2006-06-01
            ...    
401095   2011-12-14
401096   2011-09-15
401097   2011-10-28
401098   2011-08-16
401099   2011-12-14
401100   2011-08-16
401101   2011-12-14
401102   2011-08-16
401103   2011-09-15
401104   2011-08-16
401105   2011-10-25
401106   2011-08-16
401107   2011-09-15
401108   2011-08-16
401109   2011-08-16
401110   2011-09-15
401111   2011-10-25
401112   2011-10-25
401113   2011-10-25


In [0]:
??add_datepart

In [0]:
fld=df_raw.saledate

In [20]:
add_datepart(df_raw,'saledate')
df_raw.saleYear.head()

0    2006
1    2004
2    2004
3    2011
4    2009
Name: saleYear, dtype: int64

In [51]:
df_raw.columns

Index(['SalesID', 'SalePrice', 'MachineID', 'ModelID', 'datasource',
       'auctioneerID', 'YearMade', 'MachineHoursCurrentMeter', 'UsageBand',
       'fiModelDesc', 'fiBaseModel', 'fiSecondaryDesc', 'fiModelSeries',
       'fiModelDescriptor', 'ProductSize', 'fiProductClassDesc', 'state',
       'ProductGroup', 'ProductGroupDesc', 'Drive_System', 'Enclosure',
       'Forks', 'Pad_Type', 'Ride_Control', 'Stick', 'Transmission',
       'Turbocharged', 'Blade_Extension', 'Blade_Width', 'Enclosure_Type',
       'Engine_Horsepower', 'Hydraulics', 'Pushblock', 'Ripper', 'Scarifier',
       'Tip_Control', 'Tire_Size', 'Coupler', 'Coupler_System',
       'Grouser_Tracks', 'Hydraulics_Flow', 'Track_Type',
       'Undercarriage_Pad_Width', 'Stick_Length', 'Thumb', 'Pattern_Changer',
       'Grouser_Type', 'Backhoe_Mounting', 'Blade_Type', 'Travel_Controls',
       'Differential_Type', 'Steering_Controls', 'saleYear', 'saleMonth',
       'saleWeek', 'saleDay', 'saleDayofweek', 'saleDayofyear',


In [28]:
df_raw.head()

SalesID  SalePrice  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246  11.097410     999089     3157         121           3.0      2004   
1  1139248  10.950807     117657       77         121           3.0      1996   
2  1139249   9.210340     434808     7009         121           3.0      2001   
3  1139251  10.558414    1026470      332         121           3.0      2001   
4  1139253   9.305651    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter UsageBand fiModelDesc     ...     saleDay  \
0                      68.0       Low        521D     ...          16   
1                    4640.0       Low      950FII     ...          26   
2                    2838.0      High         226     ...          26   
3                    3486.0      High    PC120-6E     ...          19   
4                     722.0    Medium        S175     ...          23   

  saleDayofweek saleDayofyear saleIs_month_end saleIs_month_start  \
0             3           320            False              False   
1             4            86            False              False   
2             3            57            False              False   
3             3           139            False              False   
4             3           204            False              False   

  saleIs_quarter_end saleIs_quarter_start saleIs_year_end saleIs_year_start  \
0              False                False           False             False   
1              False                False           False             False   
2              False                False           False             False   
3              False                False           False             False   
4              False                False           False             False   

  saleElapsed  
0  1163635200  
1  1080259200  
2  1077753600  
3  1305763200  
4  1248307200  

[5 rows x 65 columns]

In [0]:
??train_cats

In [0]:
train_cats(df_raw)

In [22]:
df_raw.UsageBand.cat.categories

Index(['High', 'Low', 'Medium'], dtype='object')

In [0]:
df_raw.UsageBand.cat.set_categories(['High','Medium','Low'], ordered=True , inplace=True)

In [24]:
display_all(df_raw.isnull().sum().sort_index()/len(df_raw))

Backhoe_Mounting            0.803872
Blade_Extension             0.937129
Blade_Type                  0.800977
Blade_Width                 0.937129
Coupler                     0.466620
Coupler_System              0.891660
Differential_Type           0.826959
Drive_System                0.739829
Enclosure                   0.000810
Enclosure_Type              0.937129
Engine_Horsepower           0.937129
Forks                       0.521154
Grouser_Tracks              0.891899
Grouser_Type                0.752813
Hydraulics                  0.200823
Hydraulics_Flow             0.891899
MachineHoursCurrentMeter    0.644089
MachineID                   0.000000
ModelID                     0.000000
Pad_Type                    0.802720
Pattern_Changer             0.752651
ProductGroup                0.000000
ProductGroupDesc            0.000000
ProductSize                 0.525460
Pushblock                   0.937129
Ride_Control                0.629527
Ripper                      0.740388
S

In [0]:
os.makedirs('tmp',exist_ok=True)
df_raw.to_feather('tmp/raw')

In [34]:
df_raw=pd.read_feather('tmp/raw')

TypeError: ignored

In [0]:
df,y,nas = proc_df(df_raw,'SalePrice')

In [26]:
df.head()

SalesID  MachineID  ModelID  datasource  auctioneerID  YearMade  \
0  1139246     999089     3157         121           3.0      2004   
1  1139248     117657       77         121           3.0      1996   
2  1139249     434808     7009         121           3.0      2001   
3  1139251    1026470      332         121           3.0      2001   
4  1139253    1057373    17311         121           3.0      2007   

   MachineHoursCurrentMeter  UsageBand  fiModelDesc  fiBaseModel  \
0                      68.0          3          950          296   
1                    4640.0          3         1725          527   
2                    2838.0          1          331          110   
3                    3486.0          1         3674         1375   
4                     722.0          2         4208         1529   

              ...               saleDayofyear  saleIs_month_end  \
0             ...                         320             False   
1             ...                          86             False   
2             ...                          57             False   
3             ...                         139             False   
4             ...                         204             False   

   saleIs_month_start  saleIs_quarter_end  saleIs_quarter_start  \
0               False               False                 False   
1               False               False                 False   
2               False               False                 False   
3               False               False                 False   
4               False               False                 False   

   saleIs_year_end  saleIs_year_start  saleElapsed  auctioneerID_na  \
0            False              False   1163635200            False   
1            False              False   1080259200            False   
2            False              False   1077753600            False   
3            False              False   1305763200            False   
4            False              False   1248307200            False   

   MachineHoursCurrentMeter_na  
0                        False  
1                        False  
2                        False  
3                        False  
4                        False  

[5 rows x 66 columns]

In [0]:
??proc_df()


In [0]:
m=RandomForestRegressor(n_jobs=-1)

In [28]:
m.fit(df,y)

RandomForestRegressor(bootstrap=True, criterion='mse', max_depth=None,
           max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=-1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

In [29]:
m.score(df,y)

0.9830781663261157

In [30]:
def split_vals(a,n): return a[:n].copy(), a[n:].copy()

n_valid = 12000
n_trn=len(df)-n_valid
raw_train,raw_valid=split_vals(df_raw,n_trn)
X_train,X_valid=split_vals(df,n_trn)
y_train,y_valid=split_vals(df,n_trn)

X_train.shape, y_train.shape,X_valid.shape

((389125, 66), (389125, 66), (12000, 66))

In [0]:
def rmse(x,y): return math.sqrt((((x.astype(np.float))-(y.astype(np.float))**2).mean()))

def print_score(m):
    rmse1= rmse(m.predict(X_train), y_train)
    res = [rmse(m.predict(X_train), y_train), rmse(m.predict(X_valid), y_valid),
                m.score(X_train, y_train), m.score(X_valid, y_valid)]
    if hasattr(m, 'oob_score_'): res.append(m.oob_score_)
    print(rmse1, res)

I used np.sqrt as i got an error with math.sqrt() - The error

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-78-94bd77db69e6> in <module>()
      1 m = RandomForestRegressor(n_jobs =-1)
      2 get_ipython().magic('time m.fit(X_train,y_train)')
----> 3 print_score(m)

<ipython-input-77-871945cbb62b> in print_score(m)
      2 
      3 def print_score(m):
----> 4     rmse1= rmse(m.predict(X_train), y_train)
      5     res = [rmse(m.predict(X_train), y_train), rmse(m.predict(X_valid), y_valid),
      6                 m.score(X_train, y_train), m.score(X_valid, y_valid)]

<ipython-input-77-871945cbb62b> in rmse(x, y)
----> 1 def rmse(x,y): return math.sqrt(((x-y)**2).mean())
      2 
      3 def print_score(m):
      4     rmse1= rmse(m.predict(X_train), y_train)
      5     res = [rmse(m.predict(X_train), y_train), rmse(m.predict(X_valid), y_valid),

/usr/local/lib/python3.6/dist-packages/pandas/core/series.py in wrapper(self)
    110             return converter(self.iloc[0])
    111         raise TypeError("cannot convert the series to "
--> 112                         "{0}".format(str(converter)))
    113 
    114     return wrapper

TypeError: cannot convert the series to <class 'float'>


In [0]:
def rmse(x,y): return np.sqrt(((x-y)**2).mean())
def print_score(m):
#    rmse1= rmse(m.predict(X_train), y_train)
    res = [rmse(m.predict(X_train), y_train), rmse(m.predict(X_valid), y_valid),
                m.score(X_train, y_train), m.score(X_valid, y_valid)]
    if hasattr(m, 'oob_score_'): res.append(m.oob_score_)
    print(res)

In [38]:
m = RandomForestRegressor(n_jobs =-1)
%time m.fit(X_train,y_train)
print_score(m)

CPU times: user 6min 6s, sys: 2.18 s, total: 6min 8s
Wall time: 3min 59s
[SalesID                        16167.232432
MachineID                      16831.173108
ModelID                         2101.414677
datasource                         0.207197
auctioneerID                       4.774846
YearMade                         110.080494
MachineHoursCurrentMeter        3215.001167
UsageBand                          0.253528
fiModelDesc                      366.010106
fiBaseModel                      135.708634
fiSecondaryDesc                   13.634496
fiModelSeries                     10.350931
fiModelDescriptor                 10.571506
ProductSize                        0.712905
fiProductClassDesc                 7.074797
state                              3.589563
ProductGroup                       0.526127
ProductGroupDesc                   0.526127
Drive_System                       0.434495
Enclosure                          0.803627
Forks                              0.186664
Pa